src/devices/src/virtio/mod.rs

```diff
@@ -0,0 +1,150 @@
+    // This stands for something that implements `MmioManager`, and can be passed as a reference
+    // or smart pointer (such as a `Mutex` guard).
+    pub mmio_mgr: B,
+    // The virtio MMIO device parameters (MMIO range and interrupt to be used).
+    pub mmio_cfg: MmioConfig,
+    // We pass a mutable reference to the kernel cmdline `String` so the device can add any
+    // required arguments (i.e. for virtio over MMIO discovery). This means we need to create
+    // the devices before loading he kernel cmdline into memory, but that's not a significant
+    // limitation.
+    pub kernel_cmdline: &'a mut String,
+}
+
+/// Simple trait to model the operation of signalling the driver about used events
+/// for the specified queue.
+// TODO: Does this need renaming to be relevant for packed queues as well?
+pub trait SignalUsedQueue {
+    // TODO: Should this return an error? This failing is not really recoverable at the interface
+    // level so the expectation is the implementation handles that transparently somehow.
+    fn signal_used_queue(&self, index: u16);
+}
+
+/// Uses a single irqfd as the basis of signalling any queue (useful for the MMIO transport,
+/// where a single interrupt is shared for everything).
+pub struct SingleFdSignalQueue {
+    pub irqfd: Arc<EventFd>,
+    pub interrupt_status: Arc<AtomicU8>,
+}
+
+impl SignalUsedQueue for SingleFdSignalQueue {
+    fn signal_used_queue(&self, _index: u16) {
+        self.interrupt_status
+            .fetch_or(VIRTIO_MMIO_INT_VRING, Ordering::SeqCst);
+        self.irqfd
+            .write(1)
+            .expect("Failed write to eventfd when signalling queue");
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use vm_device::bus::MmioAddress;
+    use vm_device::device_manager::IoManager;
+    use vm_memory::{GuestAddress, GuestMemoryMmap};
+
+    use super::*;
+
+    // Can be used in other modules to test functionality that requires a `CommonArgs` struct as
+    // input. The `args` method below generates an instance of `CommonArgs` based on the members
+    // below.
+    pub struct CommonArgsContext {
+        pub mem: Arc<GuestMemoryMmap>,
+        pub vm_fd: Arc<VmFd>,
+        pub event_mgr: EventManager<Arc<Mutex<dyn MutEventSubscriber + Send>>>,
+        pub mmio_mgr: IoManager,
+        pub mmio_cfg: MmioConfig,
+        pub kernel_cmdline: String,
+    }
+
+    impl CommonArgsContext {
+        pub fn new() -> Self {
+            let mem =
+                Arc::new(GuestMemoryMmap::from_ranges(&[(GuestAddress(0), 0x1000_0000)]).unwrap());
+
+            let kvm = kvm_ioctls::Kvm::new().unwrap();
+            let vm_fd = Arc::new(kvm.create_vm().unwrap());
+
+            let range = MmioRange::new(MmioAddress(0x1_0000_0000), 0x1000).unwrap();
+            let mmio_cfg = MmioConfig { range, gsi: 5 };
+
+            // Required so the vm_fd can be used to register irqfds.
+            vm_fd.create_irq_chip().unwrap();
```

Comment:
This code will need to be duplicated in the vmm crate as well if we want to test `add_block_device`, and there we should remove this call to `create_irq_chip` because Vm::new() is expected to be called. I am wondering if it makes sense to pass the VmFd as a parameter instead, and later on have a test_utilities feature behind which we can hide this functionality.

Reply:
In the VMM case we can generate a `CommonArgs` struct backed by the vmm itself (which is needed anyway to create devices with the current model). This method is local to the devices crate to test things independently from the VMM if they ever get separated as crates. Otherwise, looks like we can just try to instantiate a VMM object as part of devices tests and generate `CommonArgs` based on that instead of mocking things.

Reply:
Instantiating a VMM is not great either because it will make devices depend on vmm and that would be a cyclic dependency. This is why I was suggesting to just pass this as a parameter as a sort of okay compromise. It is sort of a nit up to the point where we actually add aarch64 support.

Reply:
The only thing I'm wondering is whether passing that as a parameter doesn't just kick the can a bit forward, because someone still has to properly set that up. I'll implement any change, but if it's ok with you, I'd wait until we can actually test some Arm stuff as well.